README.markdown
# Personnummer

Personnummer is a ruby class that handles [Swedish personal identity number](http://en.wikipedia.org/wiki/Personal_identity_number_(Sweden)).

### Installation

    $ gem install personnummer

### Example usage

```
>> p = Personnummer.new('830428-5912')
#>>Personnummer:0x007fc971369100 @valid=true, @control_digit=2, @divider="-", @serial=591, @born=#>>PersonnummerDate:0x007fc971368318 @divider="-", @day=28, @month=4, @year=1983> #<, @region="Skaraborgs län", @female=false> #<

>> p.valid?
=> true

>> p = Personnummer.new('830428')
=> ArgumentError: The supplied personnummer is invalid
```

### This implementation also supports co-ordination numbers

[Information about co-ordination numbers (in swedish)](https://www.skatteverket.se/rattsinformation/reglerochstallningstaganden/meddelanden/2010/meddelanden2010/skvm201003.5.233f91f71260075abe8800024297.html)

These numbers are obtained by adding 60 to the day of the birth date

```
>> p = Personnummer.new('830488-5912')
=> #<Personnummer:0x007fbafd1458d8 ...>
>> p.born
=> #<Date: 1983-04-28 ((2445453j,0s,0n),+0s,2299161j)>
>> p.co_ordination_number?
=> true
```


Changelog
0.0.9 Released

  2016-09-01: Return correct standardized string for samordningsnummer

0.0.8 Released

  2014-06-29: Support for co-ordination numbers (theneubeck)

0.0.7 Released

  2012-09-12: Don't alter an existing control digit, even if it’s wrong

0.0.6 Released

  2011-04-18: Now supports century if given

0.0.5 Released

  2011-11-25: Tests!

0.0.4 Released

  2011-01-09: Ruby 1.9 support

0.0.3 Released

  2010-04-19: (Less) naïve age calculation

0.0.2 Released

  2008-08-07: Number argument can now be a Fixnum or String

0.0.1 Released


spec/spec_helper.rb
$: << File.join(File.dirname(__FILE__), '..', 'lib')

require 'cover_me'

RSpec.configure do |config|
end


spec/lib/personnummer_spec.rb
# -*- encoding: utf-8 -*-
require 'spec_helper'
require 'personnummer'
require 'timecop'

describe Personnummer do
  describe "initialize" do
    it "accepts a valid personnummer (with control digit)" do
      expect{ Personnummer.new(9001010017) }.not_to raise_error(ArgumentError)
      expect{ Personnummer.new('900101-0017') }.not_to raise_error(ArgumentError)
      expect{ Personnummer.new('100101+0017') }.not_to raise_error(ArgumentError)
      expect{ Personnummer.new('19900101-0017') }.not_to raise_error(ArgumentError)
    end

    it "accepts a valid personnummer (without control digit)" do
      expect{ Personnummer.new(900101001) }.not_to raise_error(ArgumentError)
      expect{ Personnummer.new('900101-001') }.not_to raise_error(ArgumentError)
      expect{ Personnummer.new('100101+001') }.not_to raise_error(ArgumentError)
      expect{ Personnummer.new('19900101-001') }.not_to raise_error(ArgumentError)
    end

    it "throws ArgumentError on an invalid personnummer" do
      expect{ Personnummer.new(nil) }.to raise_error(ArgumentError)
      expect{ Personnummer.new(17) }.to raise_error(ArgumentError)
      expect{ Personnummer.new('112233-4455') }.to raise_error(ArgumentError)
      expect{ Personnummer.new('19112233-4455') }.to raise_error(ArgumentError)
    end
  end

  describe "age" do
    it "calculates age correctly" do
      p = Personnummer.new('900101-001')
      p2 = Personnummer.new('19900101-001')

      Timecop.freeze(Date.parse('1980-01-01')) do
        p.age.should == 0
        p2.age.should == 0
      end

      Timecop.freeze(Date.parse('1990-01-01')) do
        p.age.should == 0
        p2.age.should == 0
      end

      Timecop.freeze(Date.parse('2000-01-01')) do
        p.age.should == 10
        p2.age.should == 10
      end

      Timecop.freeze(Date.parse('2090-01-01')) do
        p.age.should == 100
        p2.age.should == 100
      end
    end
  end

  describe "born" do
    it "returns the correct date" do
      Timecop.freeze(Date.parse('2010-01-01')) do
        Personnummer.new('900101-001').born.should == Date.parse('1990-01-01')
        Personnummer.new('010101-001').born.should == Date.parse('2001-01-01')
        Personnummer.new('100101+001').born.should == Date.parse('1810-01-01')
        Personnummer.new('19900101-001').born.should == Date.parse('1990-01-01')
        Personnummer.new('18900101-001').born.should == Date.parse('1890-01-01')
      end
    end
  end

  describe "region" do
    it "returns '' for persons born on or after 1991-01-01" do
      Personnummer.new('910101-001').region.should == ''
    end

    it "returns the correct region for persons born before 1991-01-01" do
      it_maps_range_to_region 000..139, 'Stockholms Län'
      it_maps_range_to_region 140..159, 'Uppsala län'
      it_maps_range_to_region 160..189, 'Södermanlands län'
      it_maps_range_to_region 190..239, 'Östergötlands län'
      it_maps_range_to_region 240..269, 'Jönköpings län'
      it_maps_range_to_region 270..289, 'Kronobergs län'
      it_maps_range_to_region 290..319, 'Kalmar län'
      it_maps_range_to_region 320..329, 'Gotlands län'
      it_maps_range_to_region 330..349, 'Blekinge län'
      it_maps_range_to_region 350..389, 'Kristianstads län'
      it_maps_range_to_region 390..459, 'Malmöhus län'
      it_maps_range_to_region 460..479, 'Hallands län'
      it_maps_range_to_region 480..549, 'Göteborgs och Bohus län'
      it_maps_range_to_region 550..589, 'Älvsborgs län'
      it_maps_range_to_region 590..619, 'Skaraborgs län'
      it_maps_range_to_region 620..649, 'Värmlands län'
      it_maps_range_to_region 650..659, 'Födda utomlands'
      it_maps_range_to_region 660..689, 'Örebro län'
      it_maps_range_to_region 690..709, 'Västmanlands län'
      it_maps_range_to_region 710..739, 'Kopparbergs län'
      it_maps_range_to_region 750..779, 'Gävleborgs län'
      it_maps_range_to_region 780..819, 'Västernorrlands län'
      it_maps_range_to_region 820..849, 'Jämtlands län'
      it_maps_range_to_region 850..889, 'Västerbottens län'
      it_maps_range_to_region 890..929, 'Norrbottens län'
      it_maps_range_to_region 930..999, 'Födda utomlands eller utländska medborgare födda i Sverige'
    end
  end

  describe "control_digit" do
    it "responds with the right control digit" do
      Personnummer.new(9001010017).control_digit.should == 7
      Personnummer.new('900101-0017').control_digit.should == 7
      Personnummer.new('900101-001').control_digit.should == 7
    end

    it "should not alter existing control digit, even if it is invalid" do
      p = Personnummer.new("901010-0010")
      p.to_s.should == "901010-0010"
      p.should_not be_valid
    end
  end

  describe "to_s" do
    it "uses the standard string representation" do
      Personnummer.new(9001010017).to_s.should == '900101-0017'
      Personnummer.new('900101-0017').to_s.should == '900101-0017'
      Personnummer.new('900101-001').to_s.should == '900101-0017'
      Personnummer.new('19900101-001').to_s.should == '900101-0017'
      Personnummer.new('18900101-001').to_s.should == '900101-0017'
      Personnummer.new('7010632391').to_s.should == '701063-2391' # Samordningsnummer
    end
  end

  describe "valid?" do
    it "marks a valid personnummer with control digit as valid" do
      Personnummer.new('900101-0017').should be_valid
    end

    it "marks a valid personnummer without control digit as invalid" do
      Personnummer.new('900101-001').should_not be_valid
    end

    it "throws ArgumentError on an invalid personnummer" do
      expect{ Personnummer.new('112233-4455') }.to raise_error(ArgumentError)
    end
  end

  describe "gender" do
    it "recognizes males" do
      Personnummer.new('900101-001').should be_male
      Personnummer.new('900101-003').should be_male
      Personnummer.new('900101-005').should be_male
      Personnummer.new('900101-007').should be_male
      Personnummer.new('900101-009').should be_male
    end

    it "recognizes females" do
      Personnummer.new('900101-000').should be_female
      Personnummer.new('900101-002').should be_female
      Personnummer.new('900101-004').should be_female
      Personnummer.new('900101-006').should be_female
      Personnummer.new('900101-008').should be_female
    end
  end

  describe "co-ordination number" do
    it "should accept a samordningsnummer" do
      Personnummer.new('701063-2391').should be_valid
    end
    it "should not accept an invalid samordningsnummer with less than 62" do
      Personnummer.new('900161-0017').should_not be_valid
    end
    it "should not accept an invalid samordningsnummer with more than 90" do
      Personnummer.new('900191-0017').should_not be_valid
    end
  end

  def it_maps_range_to_region(range, region)
    Personnummer.new('901231-%03d' % range.first).region.should == region
    Personnummer.new('901231-%03d' % range.last).region.should == region
  end
end


lib/personnummer_date.rb
require 'date'

class PersonnummerDate
  attr_reader :year, :month

  def initialize(century, year, month, day, divider)
    @divider = divider
    @day = day
    @month = month
    @year = calculate_century(century, year) + year
  end

  def day
    co_ordination_number? ? @day - 60 : @day
  end

  def co_ordination_number?
    @day && 60 < @day and @day < 92
  end

  def to_date
    Date.parse("#{year}-#{month}-#{day}")
  end

  def valid?
    to_date && true
  end

  def age
    if Date.today > to_date
      (Date.today - to_date).to_i/365
    else
      0
    end
  end

  private

    def calculate_century(century, year)
      # Get the current date
      today = Date.today

      if century == 0
        # Decide which century corresponds to the number
        if year < (today.year-2000) && @divider == '-'
          century = 2000
        elsif year < (today.year-2000) && @divider == '+'
          century = 1900
        elsif @divider == '+'
          century = 1800
        else
          century = 1900
        end
      else
        century *= 100
      end
      return century
    end

end

lib/personnummer.rb
# -*- encoding: utf-8 -*-
require 'date'

class Personnummer
  # Public readonly attributes
  attr_reader :region, :control_digit

  def initialize(number)
    @valid = false

    # Match the number
    number = number.to_s

    if number.match(/(\d{2}){0,1}(\d{2})(\d{2})(\d{2})([\-\+]{0,1})(\d{3})(\d{0,1})/)
      # Calculate the control digit based on the birth date and serial number
      cd = luhn_algorithm("#{$~[2]}#{$~[3]}#{$~[4]}#{$~[6]}")

      # Set control digit to calculated if it's missing
      @control_digit = $~[7].nil? || $~[7].empty? ? cd : $~[7].to_i

      # Get the different parts of the number
      century  = $~[1].to_i
      year     = $~[2].to_i
      month    = $~[3].to_i
      day      = $~[4].to_i
      @divider  = $~[5]
      @serial   = $~[6].to_i

      # Set default divider if not present
      if @divider.empty?
        @divider = '-'
      end

      # Make the personnummer valid if the checksum is correct
      @valid = true if @control_digit == cd && !$~[7].empty?

      @born  = Born.new(century, year, month, day, @divider)
      @valid = @born.valid? && @valid

      # Get the region name
      @region = region_name(@serial)

      # Check if the person is female based the serial (even == female)
      @female = (@serial % 2 == 0)
    else
      raise ArgumentError.new, "The supplied personnummer is invalid"
    end
  end

  def age
    @born.age
  end

  def born
    @born.to_date
  end

  def to_s
    born_string = born.strftime("%y%m%d")

    if co_ordination_number?
      day = born_string[4..6].to_i + 60
      born_string = [born_string[0..3], "%02d" % day].join
    end

    "%s%s%03d%d" % [born_string, @divider, @serial, @control_digit]
  end

  def valid?
    @valid
  end

  def male?
    !@female
  end

  def female?
    @female
  end

  def co_ordination_number?
    @born.co_ordination_number?
  end

private

  def luhn_algorithm(number)
    multiplications = []

    number.split(//).each_with_index do |digit, i|
      if i % 2 == 0
          multiplications << digit.to_i*2
        else
          multiplications << digit.to_i
      end
    end

    sum = 0

    multiplications.each do |num|
      num.to_s.each_byte do |character|
        sum += character.chr.to_i
      end
    end

    if sum % 10 == 0
      control_digit = 0
    else
      control_digit = (sum / 10 + 1) * 10 - sum
    end

    control_digit
  end

  def region_name(code)
    # Don't return a region name if the person was born after 1990
    # (When the previous region code was changed to a serial number)
    if @born.year > 1990
      return ''
    end

    case code
      when 000..139 then 'Stockholms Län'
      when 140..159 then 'Uppsala län'
      when 160..189 then 'Södermanlands län'
      when 190..239 then 'Östergötlands län'
      when 240..269 then 'Jönköpings län'
      when 270..289 then 'Kronobergs län'
      when 290..319 then 'Kalmar län'
      when 320..329 then 'Gotlands län'
      when 330..349 then 'Blekinge län'
      when 350..389 then 'Kristianstads län'
      when 390..459 then 'Malmöhus län'
      when 460..479 then 'Hallands län'
      when 480..549 then 'Göteborgs och Bohus län'
      when 550..589 then 'Älvsborgs län'
      when 590..619 then 'Skaraborgs län'
      when 620..649 then 'Värmlands län'
      when 650..659 then 'Födda utomlands'
      when 660..689 then 'Örebro län'
      when 690..709 then 'Västmanlands län'
      when 710..739 then 'Kopparbergs län'
      when 750..779 then 'Gävleborgs län'
      when 780..819 then 'Västernorrlands län'
      when 820..849 then 'Jämtlands län'
      when 850..889 then 'Västerbottens län'
      when 890..929 then 'Norrbottens län'
      when 930..999 then 'Födda utomlands eller utländska medborgare födda i Sverige'
    end
  end

  class Born
    attr_reader :year, :month

    def initialize(century, year, month, day, divider)
      @divider = divider
      @day     = day
      @month   = month
      @year    = calculate_century(century, year) + year
    end

    def day
      co_ordination_number? ? @day - 60 : @day
    end

    def co_ordination_number?
      @day && 60 < @day and @day < 92
    end

    def to_date
      Date.parse("#{year}-#{month}-#{day}")
    end

    def valid?
      to_date && true
    end

    def age
      if Date.today > to_date
        (Date.today - to_date).to_i/365
      else
        0
      end
    end

    private

    def calculate_century(century, year)
      # Get the current date
      today = Date.today

      if century == 0
        # Decide which century corresponds to the number
        if year < (today.year-2000) && @divider == '-'
          century = 2000
        elsif year < (today.year-2000) && @divider == '+'
          century = 1900
        elsif @divider == '+'
          century = 1800
        else
          century = 1900
        end
      else
        century *= 100
      end

      return century
    end
  end
end
